File: webapp/client/src/lib/traverse.ts
import type { Leg } from "@shared/schema";

/**
 * Convert DMS bearing to decimal degrees
 */
export function dmsToDecimal(deg: number, min: number, sec: number): number {
  return deg + min / 60 + sec / 3600;
}

/**
 * Convert decimal degrees to radians
 */
export function toRad(dd: number): number {
  return (dd * Math.PI) / 180;
}

/**
 * Convert radians to decimal degrees
 */
export function toDeg(rad: number): number {
  return (rad * 180) / Math.PI;
}

/**
 * Format decimal degrees to DMS string
 */
export function decimalToDMS(dd: number): string {
  const d = Math.floor(dd);
  const mf = (dd - d) * 60;
  let m = Math.floor(mf);
  let s = Math.round((mf - m) * 60);
  // handle carry-over when seconds round up to 60
  if (s >= 60) { s = 0; m += 1; }
  if (m >= 60) { m = 0; }
  return `${d}°${String(m).padStart(2, "0")}'${String(s).padStart(2, "0")}"`;
}

/**
 * Format a whole-circle bearing (0-360) as DMS
 */
export function bearingToDMS(bearingDD: number): string {
  const norm = ((bearingDD % 360) + 360) % 360;
  return decimalToDMS(norm);
}

/**
 * Compute traverse misclose and precision
 */
export interface TraverseResult {
  // Raw coordinate sums
  sumEasting: number;   // ΣΔE
  sumNorthing: number;  // ΣΔN
  // Misclose vector
  miscloseDist: number;  // linear misclose (m)
  miscloseDir: number;   // bearing of misclose vector (decimal degrees)
  miscloseDistStr: string;
  // (duplicate removed)
  miscloseDirectionStr: string;
  // Precision
  totalDist: number;
  precision: number;     // 1:X ratio
  precisionStr: string;
  // Individual coordinate differences
  coords: { easting: number; northing: number; bearingDD: number }[];
  // Quality rating
  quality: "excellent" | "good" | "fair" | "poor";
}

export function computeTraverse(legs: Leg[]): TraverseResult | null {
  if (legs.length < 2) return null;

  const coords: { easting: number; northing: number; bearingDD: number }[] = [];
  let sumE = 0;
  let sumN = 0;
  let totalDist = 0;

  for (const leg of legs) {
    const bearingDD = dmsToDecimal(leg.bearingDeg, leg.bearingMin, leg.bearingSec);
    const rad = toRad(bearingDD);
    const dE = leg.distance * Math.sin(rad);
    const dN = leg.distance * Math.cos(rad);
    coords.push({ easting: dE, northing: dN, bearingDD });
    sumE += dE;
    sumN += dN;
    totalDist += leg.distance;
  }

  const miscloseDist = Math.sqrt(sumE * sumE + sumN * sumN);
  // bearing of misclose vector
  let miscloseDir = toDeg(Math.atan2(sumE, sumN));
  miscloseDir = ((miscloseDir % 360) + 360) % 360;

  const precision = miscloseDist > 0 ? Math.round(totalDist / miscloseDist) : Infinity;

  // Quality thresholds (common surveying standards)
  let quality: TraverseResult["quality"];
  if (precision === Infinity || precision >= 10000) quality = "excellent";
  else if (precision >= 5000) quality = "good";
  else if (precision >= 2000) quality = "fair";
  else quality = "poor";

  return {
    sumEasting: sumE,
    sumNorthing: sumN,
    miscloseDist,
    miscloseDir,
    miscloseDistStr: `${miscloseDist.toFixed(4)} m`,
    miscloseDirectionStr: bearingToDMS(miscloseDir),
    totalDist,
    precision,
    precisionStr: precision === Infinity ? "∞" : `1 : ${precision.toLocaleString()}`,
    coords,
    quality,
  };
}

/**
 * Parse a bearing string like "123.4530" or "123°45'30"" or "123 45 30"
 * Returns { deg, min, sec } or null if invalid
 */
export function parseBearingString(input: string): { deg: number; min: number; sec: number } | null {
  // Try DMS with separators: "123°45'30.5"" or "123 45 30.5" or "123-45-30.5"
  const dmsMatch = input.trim().match(/^(\d{1,3})[°\-\s](\d{1,2})['\-\s](\d{1,2}(?:\.\d+)?)"?$/);
  if (dmsMatch) {
    const deg = parseInt(dmsMatch[1]);
    const min = parseInt(dmsMatch[2]);
    const sec = parseFloat(dmsMatch[3]);
    if (deg <= 359 && min <= 59 && sec < 60) return { deg, min, sec };
  }

  // Try pure decimal degrees "123.456"
  const ddMatch = input.trim().match(/^(\d{1,3})\.(\d+)$/);
  if (ddMatch) {
    const full = parseFloat(input);
    if (full < 360) {
      const deg = Math.floor(full);
      const mf = (full - deg) * 100;  // assume DDMMSS.S compact format
      const min = Math.floor(mf);
      const sec = (mf - min) * 100;
      if (min <= 59 && sec < 60) return { deg, min, sec };
      // else treat as fractional degrees
      const min2 = Math.floor((full - deg) * 60);
      const sec2 = ((full - deg) * 60 - min2) * 60;
      return { deg, min: min2, sec: sec2 };
    }
  }

  // Try compact DDDMMSS "1234530" style
  const compactMatch = input.trim().match(/^(\d{1,3})(\d{2})(\d{2}(?:\.\d+)?)$/);
  if (compactMatch) {
    const deg = parseInt(compactMatch[1]);
    const min = parseInt(compactMatch[2]);
    const sec = parseFloat(compactMatch[3]);
    if (deg <= 359 && min <= 59 && sec < 60) return { deg, min, sec };
  }

  return null;
}

export function formatBearingDMS(deg: number, min: number, sec: number): string {
  const s = Math.round(sec);
  return `${String(deg).padStart(3, "0")}°${String(min).padStart(2, "0")}'${String(s).padStart(2, "0")}"`;
}

File: survey_cogo/traverse.py
"""
survey_cogo.traverse
--------------------
Closed traverse misclose and precision computation.

A traverse is a series of lines defined by a WCB bearing and horizontal
distance. For a closed traverse the sum of ΔE and ΔN should return to
zero; any residual is the misclose vector.

Precision ratio  =  Total perimeter ÷ Linear misclose  (expressed 1 : X)

Quality thresholds (common Australian surveying standards):
  Excellent  ≥ 1 : 10 000
  Good       ≥ 1 :  5 000
  Fair       ≥ 1 :  2 000
  Poor         <  1 :  2 000
"""

import math
from .bearings import dms_to_dd, fmt_bearing

__all__ = ["compute_traverse", "print_traverse"]


def compute_traverse(legs: list[dict]) -> dict:
    """
    Compute misclose and precision for a closed traverse.

    Parameters
    ----------
    legs : list of dict
        Each dict must contain:
          'bearing_deg' (int)   — degrees  0–359
          'bearing_min' (int)   — minutes  0–59
          'bearing_sec' (int)   — seconds  0–59
          'distance'    (float) — horizontal distance in metres
        Optional:
          'label' (str)         — line identifier (e.g. "Line 1", "AB")

    Returns
    -------
    dict
        sum_e            (float)  — ΣΔE in metres
        sum_n            (float)  — ΣΔN in metres
        misclose_dist    (float)  — linear misclose in metres
        misclose_dir     (float)  — bearing of misclose vector (decimal °)
        misclose_dir_str (str)    — misclose bearing as DMS string
        total_dist       (float)  — total perimeter distance in metres
        precision        (int | float)  — 1:X value (float('inf') if perfect)
        precision_str    (str)    — e.g. "1 : 12 500"
        quality          (str)    — "excellent" | "good" | "fair" | "poor"
        lines            (list)   — per-line computation details

    Raises
    ------
    ValueError — if fewer than 2 legs are supplied

    Examples
    --------
    >>> legs = [
    ...     {"bearing_deg": 45,  "bearing_min": 30, "bearing_sec": 0, "distance": 100.0},
    ...     {"bearing_deg": 135, "bearing_min": 30, "bearing_sec": 0, "distance": 100.0},
    ...     {"bearing_deg": 225, "bearing_min": 30, "bearing_sec": 0, "distance": 100.0},
    ...     {"bearing_deg": 315, "bearing_min": 30, "bearing_sec": 0, "distance": 100.0},
    ... ]
    >>> result = compute_traverse(legs)
    >>> result["quality"]
    'excellent'
    """
    if len(legs) < 2:
        raise ValueError("A traverse requires at least 2 lines.")

    lines      = []
    sum_e      = 0.0
    sum_n      = 0.0
    total_dist = 0.0

    for leg in legs:
        bearing_dd = dms_to_dd(
            leg["bearing_deg"],
            leg["bearing_min"],
            leg["bearing_sec"],
        )
        rad  = math.radians(bearing_dd)
        de   = leg["distance"] * math.sin(rad)
        dn   = leg["distance"] * math.cos(rad)
        sum_e      += de
        sum_n      += dn
        total_dist += leg["distance"]

        lines.append({
            "label":       leg.get("label", ""),
            "bearing_dd":  bearing_dd,
            "bearing_str": fmt_bearing(bearing_dd),
            "distance":    leg["distance"],
            "delta_e":     de,
            "delta_n":     dn,
        })

    misclose_dist = math.sqrt(sum_e ** 2 + sum_n ** 2)
    misclose_dir  = (math.degrees(math.atan2(sum_e, sum_n)) % 360 + 360) % 360

    if misclose_dist > 1e-9:
        precision = round(total_dist / misclose_dist)
    else:
        precision = float("inf")

    if precision == float("inf") or precision >= 10_000:
        quality = "excellent"
    elif precision >= 5_000:
        quality = "good"
    elif precision >= 2_000:
        quality = "fair"
    else:
        quality = "poor"

    prec_str = "∞" if precision == float("inf") else f"1 : {int(precision):,}"

    return {
        "sum_e":            sum_e,
        "sum_n":            sum_n,
        "misclose_dist":    misclose_dist,
        "misclose_dir":     misclose_dir,
        "misclose_dir_str": fmt_bearing(misclose_dir),
        "total_dist":       total_dist,
        "precision":        precision,
        "precision_str":    prec_str,
        "quality":          quality,
        "lines":            lines,
    }


def print_traverse(result: dict, name: str = "Traverse") -> None:
    """
    Pretty-print traverse results to stdout.

    Parameters
    ----------
    result : dict — return value of compute_traverse()
    name   : str  — traverse name shown in header
    """
    w = 72
    print("\n" + "=" * w)
    print(f"  TRAVERSE — {name}")
    print("=" * w)
    print(f"  {'Line':<18} {'Bearing':>12}  {'Dist (m)':>10}"
          f"  {'ΔE (m)':>10}  {'ΔN (m)':>10}")
    print("  " + "-" * (w - 2))

    for i, ln in enumerate(result["lines"], 1):
        lbl = ln["label"] or f"Line {i}"
        print(f"  {lbl:<18} {ln['bearing_str']:>12}"
              f"  {ln['distance']:>10.3f}"
              f"  {ln['delta_e']:>+10.4f}"
              f"  {ln['delta_n']:>+10.4f}")

    print("  " + "-" * (w - 2))
    print(f"  {'Total distance':<36} {result['total_dist']:>10.3f} m")
    print(f"  {'ΣΔE':<36} {result['sum_e']:>+10.4f} m")
    print(f"  {'ΣΔN':<36} {result['sum_n']:>+10.4f} m")
    print(f"  {'Linear misclose':<36} {result['misclose_dist']:>10.4f} m")
    print(f"  {'Direction of misclose':<36} {result['misclose_dir_str']:>12}")
    print(f"  {'Precision':<36} {result['precision_str']:>12}")
    print(f"  {'Quality':<36} {result['quality'].upper():>12}")
    print("=" * w)
